File: flutter/lib/domain/use_case/ai/generate_deck_with_ai_use_case.dart
import 'package:flashcard/domain/entities/flashcard.dart';
import 'package:flashcard/domain/repository/ai/ai_generator_repository.dart';

class GenerateDeckWithAIUseCase {
  final AIGeneratorRepository _aiRepository;

  GenerateDeckWithAIUseCase(
    this._aiRepository
  );

  Future<Map<String, Flashcard>> call({
    String? deckId,
    int count = 5,
    String prompt = ""}
      ) async {
    final generatedFlashcards = await _aiRepository.generateDeck(
        deckId,
        count,
        prompt
    );
    return generatedFlashcards;
  }
}

File: android/app/src/main/java/com/synergyboat/flutterai/domain/usecase/ai/GenerateDeckWithAIUseCase.kt
package com.synergyboat.flutterai.domain.usecase.ai

import com.synergyboat.flutterai.domain.entities.Deck
import com.synergyboat.flutterai.domain.repository.ai.AiGeneratorRepository
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json

class GenerateDeckWithAIUseCase(
    private val repository: AiGeneratorRepository
) {
    suspend operator fun invoke(
        deckId: Int? = null,
        count: Int = 10,
        prompt: String = ""
    ): Deck {
        val response: String? = repository.generateDeck(deckId = deckId, count = count, prompt = prompt)

        // You can replace this with proper serialization logic
        return if (response != null) {
            decodeDeckFromJson(response)
        } else {
            throw IllegalStateException("AI generation failed: null response")
        }
    }

    private fun decodeDeckFromJson(json: String): Deck {
        return Json.decodeFromString(Deck.serializer(),json)
    }
}